CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format

veronikaslc opened the pull request:
Added more filters: Created by, created date, modified date and status.

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Added more filters

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -342,6 +441,108 @@ function ExportButton(props) {
+                  >
+                    { users?.map(v =>
+                            <MenuItem value={v.name} key={`option-${v.principalName}`} className={classes.variableOption}>
+                              <Link to={v.path} underline="hover">{v.principalName}</Link>
```
`Link` from mui doesn't have a `to` prop, that's a prop of the `react-router-dom` Link Also, there's no reason to have links here, simply the name of the user will suffice.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -113,11 +119,23 @@ function ExportButton(props) {
   // Column selection:
   // just one of the Include or Exclude options should be available at a time
-  const [ isInclude, setIsInclude ] = useState(false);
+  const [ isIncludeEntities, setIsIncludeEntities ] = useState(false);
```
I find this very hard to read as a variable name.

veronikaslc replied:
Do you want me to rename it back?

marta- replied:
Renamed / refactored in the base branch

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -342,6 +441,108 @@ function ExportButton(props) {
+              </Grid>
+              <Grid item xs={8}>
+
+                {/* List the entered values */}
+                { statusIds.map((value, index) =>
+                  <Grid container
+                    key={`${value}-${index}`}
+                    direction="row"
+                    justifyContent="space-between"
+                    alignItems="stretch"
+                    className={classes.valueEntry}
+                  >
+                    <Grid item xs={9}>{ value }</Grid>
+                    <Grid item xs={3} className={classes.valueActions}>
+                      <Tooltip title="Delete status">
+                        <IconButton onClick={() => deleteStatus(statusIds.indexOf(value.path))}><CloseIcon/></IconButton>
+                      </Tooltip>
+                    </Grid>
+                  </Grid>
+                )}
+
+                <FormControl variant="standard" fullWidth className={classes.variableDropdown}>
+                  <InputLabel id="status">Select statuses</InputLabel>
```
The backend doesn't support more than one status at a time. The status control should be single select, and the label shouls read "Select a status".

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -210,6 +289,28 @@ function ExportButton(props) {
   }
+
+  let getDatePicker = (value, setter, label) => {
+    return (<DatePicker            
```
Unnecessary trailing whitespace:
```suggestion
    return (<DatePicker
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -342,6 +410,89 @@ function ExportButton(props) {
+
+          <Grid container alignItems='center' direction="row" className={classes.container}>
+            <Grid item xs={4}>
+                <Typography variant="subtitle2">{isIncludeStatus ? "Statuses to include:" : "Statuses to exclude:"}</Typography>
```
```suggestion
                <Typography variant="subtitle2">{isIncludeStatus ? "Status to include:" : "Status to exclude:"}</Typography>
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -342,6 +410,89 @@ function ExportButton(props) {
+          </Grid>
+
+          <Grid container alignItems='center' direction="row" className={classes.container}>
+              <Grid item xs={4}><Typography variant="subtitle2">Statuses selection mode:</Typography></Grid>
```
```suggestion
              <Grid item xs={4}><Typography variant="subtitle2">Status selection mode:</Typography></Grid>
```

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Refactoring: better styling for date range filters

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Status flag filter: refactoring and label improvements

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
"Created by" filter: small alignment improvement, removed unneeded classNames and wrong label

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Indentation cleanup

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -215,6 +269,24 @@ function ExportButton(props) {
+                label={label}
+                value={value}
+                onChange={(value) => {
+                  setter(value);
```
It would be nice to have some date range validation

veronikaslc replied:
First I implemented it, and then I though: What if the user wants to export not all forms in between the dates but the opposite - all forms outside of the range? Shell we support such thing? Can the backend do that? Do we need to restrict the export functionality for only a time range "from-to"?

marta- replied:
This is how the backend translates these filters into a query: https://github.com/data-team-uhn/cards/blob/5bd1edb22a32dad7b264eea6f7d009d72fee9c3b/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/DataProcessor.java#L232-L254

As can be seen here, the conditions are `and`-ed, which means that if the "after" and "before" limits are not in order then no results will ever be found.

veronikaslc replied:
Do we want to add `if` with `or-` for the cases when limits are not in order to support out-of-range export? Or the cases it is useful are way too few?

sdumitriu replied:
We should keep it simple for now. The user can export two different CSVs and paste them together if needed, but I doubt this is plausible requirement.

veronikaslc replied:
Fixed in [ef5d2cf](https://github.com/data-team-uhn/cards/pull/1395/commits/ef5d2cfef43648d0c7cb6ce01b68106ec93989ed)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -215,6 +269,24 @@ function ExportButton(props) {
   }
+
+  let getDatePicker = (value, setter, label) => {
+    return (<DatePicker
+                label={label}
+                value={value}
+                onChange={(value) => {
+                  setter(value);
+                }}
+                renderInput={(params) =>
+                  <TextField
+                    variant="standard"
+                    helperText={null}
+                    {...params}
+                  />
+                }
+              />);
```
In the ui the dates are displayed as without a time, but when converted to filters they include time. Please enable time in the ui to resolve the inconsistency.

veronikaslc replied:
Sure, it's just the specs confused me with `Modified: after <date> - before <date>` so I thought we only need a date without time. Easy change

marta- replied:
In that case the time should have been stripped when generating the filter's value. To be fair, the most common use case is indeed to pass a date with "midnight" time, which is why "date" was used in the specs. The problem is not so much the lack of time setting on the picker, but more the fact that what you see (pure date) is not what you get (results filtered by that date with some time setting that you have no control over).

veronikaslc replied:
Now stripping the time when geenrating the filters values in [49f3112](https://github.com/data-team-uhn/cards/pull/1395/commits/49f3112b1dd1145722f2f88e1f7f9eda487f13eb)

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
strip the time when generating the filter's value

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Added dates range validation

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -280,8 +291,15 @@ function ExportButton(props) {
                     variant="standard"
-                    helperText={null}
                     {...params}
+                    error={label == "before" ? title == "created" ? createdRangeIsInvalid : modifiedRangeIsInvalid : false}
+                    helperText={label == "before" ?
+                                 (title == "created" && createdRangeIsInvalid) || (title == "modified" && modifiedRangeIsInvalid)
+                                 ? "The before date should be later than after date." : "" 
+                                 : ""}
+                    InputProps={{
+                      ...params.InputProps
+                    }}
```
Do we really need the `title`? can't we just pass the appropriate flag (`createdRangeIsInvalid` or `modifiedRangeIsInvalid`) as a `rangeIsInvalid` parameter?

marta- replied:
Done in 7eb38ab

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Simplify date range validation logic

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
When there is a date range error, mark both dates in the range and place the error message under the entire range, not just the "before" date

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Fixed Grid container nesting and some indentation

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Separate the export config form into sections for better readability

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Better alignment for entries with tall select controls

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
* Make date range controls more user friendly (the "after" and "before" labels previously used can be confusing)
* Refactoring to reduce code duplication when rendering date ranges

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -146,6 +184,25 @@ function ExportButton(props) {
+
+  // Determine if the before date is earlier than the after date
+  useEffect(() => {
+    open && setCreatedRangeIsInvalid(!!createdAfter && !!createdBefore && new Date(createdBefore).valueOf() < new Date(createdAfter).valueOf());
```
The comparison should be done with `<=` instead of `<`. In the backend, the "created" date is compared like this to the filter:

https://github.com/data-team-uhn/cards/blob/5bd1edb22a32dad7b264eea6f7d009d72fee9c3b/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/serialize/DataProcessor.java#L234-L239

which means the two values must actually be different in order to have a chance at getting any results.

Another issue is that this compares the dates selected (full date down to milliseconds) not the dates passed to the dataFilter. That means if I select the same year/month/day in both 3 seconds apart, the range will pass the validity test but the filter will never return any results because the dates we trim down to are identical.

This second issue can be avoided if we don't restrict the filters to date only, and instead use the full date-time.

veronikaslc replied:
Fixed in [e091901](https://github.com/data-team-uhn/cards/pull/1395/commits/e091901eb9a0908383154d71296e9bde0d3f243a)

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -174,6 +231,26 @@ function ExportButton(props) {
+      path += ".dataFilter:createdBy=" + user;
+    }
+    if (createdAfter) {
+      path += ".dataFilter:createdAfter=" + createdAfter.toISODate();
```
As suggested in the previous comment, let's send the full date-time instead of trimming down to year/month/day, to make sure the frontend can support all that the backend is capable of.

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -215,6 +292,49 @@ function ExportButton(props) {
+
+  let getDatePicker = (value, setter, rangeIsInvalid) => {
+    return (<LocalizationProvider dateAdapter={AdapterLuxon}>
+              <DatePicker
```
Per previous comments, a DateTime picker should be used here.

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Replaced DatePicker with DateTimePicker for full functionality control

marta- commented:
The American date format (month/day/year) will never cease to confuse me. Let's use a less ambiguous one, `year/month/day time`.
![image](https://user-images.githubusercontent.com/651980/223558121-a4f90f7c-9204-4100-8b08-8ec43746a5e2.png)

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format
Change date format to start with dd/MM/yyyy

veronikaslc committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format

yyyy/MM/dd date format

marta- reviewed (approved):
The code and the user experience look good to me. I only tested the UI, didn't do actual filtering tests. The generated URL appears right to me, but someone needs to at least test that filtering by date works as expected @sashaandjic

marta- commented:
Rebased on the latest `dev` so this can be tested post react upgrade.

sashaandjic commented:
Label `Don't know / not sure` looks odd in export.
![image](https://user-images.githubusercontent.com/23243244/223609665-79f8f3cf-52c6-4cb2-af38-4aa2e834447f.png)

sashaandjic commented:
The filtering by date works as expected.

marta- commented:
> Label `Don't know / not sure` looks odd in export.

It's not related to these changes and should not prevent the merge if there are no other issues.

I thought https://github.com/data-team-uhn/cards/pull/969/commits/69263fe918b3991b6792efcc143bff68179a6293 should have fixed it a long time ago. I've made a quick fix PR to replace the special characters in case we don't find a better solution: #1409

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/dataHomepage/ExportButton.jsx:
```diff
@@ -174,6 +231,26 @@ function ExportButton(props) {
+      path += ".dataFilter:modifiedBefore=" + modifiedBefore.toISO();
+    }
+    if (status) {
+      path += ".dataFilter";
```
Why this line?

marta- replied:
If I had to guess, I would assume it's because lines 225-230 were copied over and adapted:
```
 if (selectedEntityIds.length > 0) {
      path +=  ".questionnaireFilter";
      let pref = `.questionnaireFilter:${columnSelectionMode}=`;
      for (let id in selectedEntityIds) {
        path += pref + encodeURIComponent(encodeURIComponent(selectedEntityIds[id]));
      }
    }
```

Indeed it's not needed.
```suggestion
```

marta- replied:
Removed in 7a31a88

sdumitriu reviewed (approved):
Looks very nice.

marta- committed:
CARDS-1709: Enable administrators to export all patient data stored in specific questionnaire in csv format

Remove unnecessary selector
